### PR TITLE
Hardware crypto support

### PR DESCRIPTION
**Pull request by oleorhagen**
Implementation of the HSM support for the cpp client

### CODE REVIEW

File: common/crypto.hpp [hunk 1]
```diff
@@ -46,6 +46,10 @@ class PrivateKey {
 public:
 	PrivateKey() {};
 
+	static ExpectedPrivateKey Load(const string &private_key_path) {
+		return Load(private_key_path, "");
+	}
```
Comment (lluiscampos) [inline common/crypto.hpp]: In the `.cpp` file, we have a similar wrapper:
```
ExpectedPrivateKey PrivateKey::LoadFromPEM(const string &private_key_path) {
	return PrivateKey::LoadFromPEM(private_key_path, "");
}
```

Consider moving the definition of `Load` to the cpp or to bring `LoadFromPEM` to the header :pray:

File: common/crypto.hpp [hunk 2]
```diff
@@ -15,6 +15,8 @@
 #ifndef MENDER_COMMON_CRYPTO_HPP
 #define MENDER_COMMON_CRYPTO_HPP
 
+#include <common/crypto/platform/openssl/openssl_config.h>
```
Comment (lluiscampos) [inline common/crypto.hpp:18]: ~This cannot be right: the platform agnostic header shouldn't know about openssl.~

Edit: too bad that we need this include for the definitions below. At least it should be guarded with `ifdef MENDER_CRYPTO_OPENSSL`

File: common/crypto.hpp [hunk 3]
```diff
@@ -39,34 +41,67 @@ enum CryptoErrorCode {
 	VerificationError,
 };
 
+struct Args {
+	const string private_key_path;
+	const string private_key_passphrase;
+	const string ssl_engine;
+};
```
Comment (lluiscampos) [inline common/crypto.hpp:48]: Good thinking!

File: common/crypto.hpp [hunk 4]
```diff
@@ -39,34 +41,67 @@ enum CryptoErrorCode {
...
+		key(std::move(private_key)),
+		engine {std::move(engine)} {};
+#else
+	PrivateKey(
+		unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)> &&private_key,
+		unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)> &&default_provider,
+		unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)> &&hsm_provider) :
+		key(std::move(private_key)),
+		default_provider {std::move(default_provider)},
+		hsm_provider {std::move(hsm_provider)} {};
```
Comment (lluiscampos) [inline common/crypto.hpp:103]: Can we have some alias around?
```
using PkeyPtr = unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)>;
using ProviderPtr = unique_ptr<OSSL_PROVIDER, int (*)(OSSL_PROVIDER *)>;
...
```

For being a header file, it became quite hard to read :eye:

File: common/crypto/platform/openssl/crypto.cpp [hunk 5]
```diff
@@ -148,8 +160,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
...
-	return PrivateKey::LoadFromPEM(private_key_path, "");
+#ifdef MENDER_CRYPTO_OPENSSL_LEGACY
+ExpectedPrivateKey PrivateKey::LoadFromHSM(const Args &args) {
+	log::Trace("Loading the private key from HSM");
+
+	ENGINE_load_builtin_engines();
+	auto engine = unique_ptr<ENGINE, void (*)(ENGINE *)>(
+		ENGINE_by_id(args.ssl_engine.c_str()), engine_free_func);
+
+	if (engine == NULL) {
```
Comment (lluiscampos) [inline common/crypto/platform/openssl/crypto.cpp:171]: Elsewhere we are using `nullptr`

File: common/crypto/platform/openssl/crypto.cpp [hunk 6]
```diff
@@ -148,8 +160,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
...
+				"Unhandled OpenSSL type: expected PrivateKey, got: " + info_type_string));
+			break;
+		}
+	}
+
+	return expected::unexpected(MakeError(
+		SetupError,
+		"Failed to load the private key from the hardware security module: "
+			+ GetOpenSSLErrorMessage()));
+}
```
Comment (lluiscampos) [inline common/crypto/platform/openssl/crypto.cpp:283]: It is really hard to follow. I am glad we have a new OpenSSL expert in the team :sweat_smile: 

I browsed the man pages and your code makes sense, but I am not able to provide any better feedback than that.

File: mender-auth/cli/actions.cpp [hunk 7]
```diff
@@ -45,21 +45,12 @@ shared_ptr<MenderKeyStore> KeystoreFromConfig(
...
 		pem_file = config.security.auth_private_key;
 		ssl_engine = config.security.ssl_engine;
 		static_key = cli::StaticKey::Yes;
-	}
-	if (config.https_client.key == "" && config.security.auth_private_key == "") {
+	} else {
 		pem_file = config.paths.GetKeyFile();
-		ssl_engine = config.https_client.ssl_engine;
 		static_key = cli::StaticKey::No;
 	}
```
Comment (lluiscampos) [inline mender-auth/cli/actions.cpp:55]: I think it is not right to associate `StaticKey::No` to everything but `config.security.auth_private_key != ""`. In other words, this sets no static keys always for the https configuration, potentially working incorrectly with hardware crypto for example (it will work fine as-is, because we don't have code to attempt regenerating the https key anyway, but still the design is wrong I think).

Sorry for opening now this melon, but I think this should be a filed exposed in the configuration. It is just too much of a guess to figure this out for the user.

Comment (lluiscampos) [reply mender-auth/cli/actions.cpp:55]: By the way, thank you for addressing this! :muscle:

**Review by lluiscampos — COMMENTED** [review]
Brave work!

File: common/crypto.hpp (hunk 3 above)
Comment (oleorhagen) [reply common/crypto.hpp:48]: I am also pondering typing some of these, so that it will be harder to mess up the argument order, but might be I'm going overboard here...

File: common/crypto.hpp (hunk 2 above)
Comment (oleorhagen) [reply common/crypto.hpp:18]: Yeh, issue is the `PrivateKey` needs to keep the `HSM` resources alive, so the `ENGINE` or the `PROVIDER`.

I have been thinking about how to abstract this away, but I can't think of any good ways atm.

All open to suggestions here ofc

File: common/crypto/platform/openssl/crypto.cpp (hunk 6 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:283]: It's a shame, I have been trying to make this as readable as possible. I could perhaps comment it out more?

File: mender-auth/cli/actions.cpp (hunk 7 above)
Comment (oleorhagen) [reply mender-auth/cli/actions.cpp:55]: Discussed f2f, keeping it as it is.

File: common/crypto.hpp (hunk 2 above)
Comment (kacf) [reply common/crypto.hpp:18]: One thing which occurred to me after our talk: Do the types need to be qualified? They are just pointers, so can't they be forward declared with `class ENGINE;` and that should be enough?

File: mender-auth/cli/keystore.cpp [hunk 8]
```diff
@@ -70,11 +70,11 @@ error::Error MenderKeyStore::Load() {
 }
 
 error::Error MenderKeyStore::Save() {
-	if (!key_) {
+	if (!key_.Get()) {
```
Comment (kacf) [inline mender-auth/cli/keystore.cpp:74]: Not wrong, but maybe it's more idiomatic to provide

```
operator bool() {
    return Get() != nullptr;
}
```

in PrivateKey?

File: common/crypto.hpp (hunk 4 above)
Comment (kacf) [reply common/crypto.hpp:103]: +1!

This should also help on reading the code where you use the constructors below.

File: common/crypto/platform/openssl/crypto.cpp [hunk 9]
```diff
@@ -74,6 +80,11 @@ auto bn_free = [](BIGNUM *bn) {
 		BN_free(bn);
 	}
 };
+auto engine_free_func = [](ENGINE *e) {
```
Comment (kacf) [inline common/crypto/platform/openssl/crypto.cpp:83]: Nitpick: Regular function syntax here, maybe?

File: common/crypto/platform/openssl/crypto.cpp [hunk 10]
```diff
@@ -148,8 +160,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
...
+	}
+
+	// Go through all objects in the context till we find the first private key
+	while (not OSSL_STORE_eof(ctx.get())) {
+		OSSL_STORE_INFO *info = OSSL_STORE_load(ctx.get());
+
+		if (info == NULL) {
+			return expected::unexpected(MakeError(
+				SetupError,
+				"Failed to read the private key information the hardware security module: "
```
Comment (kacf) [inline common/crypto/platform/openssl/crypto.cpp:249]: "...**from** the hardware security module"

File: common/crypto/platform/openssl/crypto.cpp [hunk 11]
```diff
@@ -148,8 +160,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
...
+			"OSSL_STORE_OPEN: Failed to load the private key from the hardware security module: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	// Go through all objects in the context till we find the first private key
+	while (not OSSL_STORE_eof(ctx.get())) {
+		OSSL_STORE_INFO *info = OSSL_STORE_load(ctx.get());
+
+		if (info == NULL) {
+			return expected::unexpected(MakeError(
```
Comment (kacf) [inline common/crypto/platform/openssl/crypto.cpp:247]: Or should we try the next key here with `continue`?

This goes for several `returns` below as well.

File: common/http/platform/beast/http.cpp [hunk 12]
```diff
@@ -270,11 +272,19 @@ error::Error Client::Initialize() {
...
+				return exp_key.error().WithContext(
+					"Error loading private key from " + client_config_.client_cert_key_path);
+			}
+
+			const int ret =
+				SSL_CTX_use_PrivateKey(ssl_ctx_[i].native_handle(), exp_key.value().Get());
+			if (ret != 1) {
+				log::Error(
+					"Failed to add the PrivateKey: " + client_config_.client_cert_key_path
+					+ " to the SSL CTX. The HTTP client will not be functioning normally");
```
Comment (kacf) [inline common/http/platform/beast/http.cpp:287]: I think we should return error here as well, we have opted for returns on all other error paths.

File: mender-auth/cli/actions.cpp [hunk 13]
```diff
@@ -97,7 +88,12 @@ error::Error DoAuthenticate(context::MenderContext &main_context) {
 	auto err = auth_client::FetchJWTToken(
 		client,
 		config.servers,
-		config.paths.GetKeyFile(),
+		{
+			config.security.auth_private_key == "" ? config.paths.GetKeyFile()
+												   : config.security.auth_private_key,
+			"",
```
Comment (kacf) [inline mender-auth/cli/actions.cpp:94]: What about passphrase?

File: mender-auth/cli/actions.cpp [hunk 14]
```diff
@@ -151,8 +147,10 @@ error::Error DaemonAction::Execute(context::MenderContext &main_context) {
 	ipc::Server ipc_server {loop, config};
 
 	err = ipc_server.Listen(
-		config.security.auth_private_key == "" ? config.paths.GetKeyFile()
-											   : config.security.auth_private_key,
+		{config.security.auth_private_key == "" ? config.paths.GetKeyFile()
+												: config.security.auth_private_key,
+		 "",
```
Comment (kacf) [inline mender-auth/cli/actions.cpp:152]: What about passphrase?

File: mender-auth/cli/keystore.cpp (hunk 8 above)
Comment (oleorhagen) [reply mender-auth/cli/keystore.cpp:74]: Good suggestion

File: common/crypto.hpp (hunk 2 above)
Comment (oleorhagen) [reply common/crypto.hpp:18]: I was pondering that all evening last night!

I'm not sure if it's a good idea though, is it considered kosher?

File: mender-auth/cli/actions.cpp (hunk 13 above)
Comment (oleorhagen) [reply mender-auth/cli/actions.cpp:94]: Yes, glad you pointed it out.

Not sure what to do about it atm. Do we currently have any support for it? I can't find it.

This could be a good opportunity for adding it though

File: common/crypto.hpp (hunk 2 above)
Comment (kacf) [reply common/crypto.hpp:18]: Yes, this is quite common, actually. It doesn't apply to us, but in public library development there is an important function to doing so. If your private struct is exposed in the header, there is a chance it could, maybe unintentionally, be used directly in the source files of whoever is using your library. This is bad, because it binds the memory layout of that struct to the binary that was built, instead of to the library. This means you can no longer update the struct, because you risk breaking compatibility with already released binaries. OTOH, if you only forward declare it, then any such unintentional use would be flagged by the compiler as illegal, and the users would be forced to use the factory functions you provide to get a pointer to the struct.

File: mender-auth/cli/actions.cpp (hunk 13 above)
Comment (kacf) [reply mender-auth/cli/actions.cpp:94]: It's being set [here](https://github.com/mendersoftware/mender/pull/1429/files#diff-3540d3121deba3ec9ef7262bf0093173c7717d1bfc172970cfe1a527644ac9e4R75) by command line options.

Comment (oleorhagen) [reply mender-auth/cli/actions.cpp:94]: lol, of course!

File: common/crypto/platform/openssl/crypto.cpp (hunk 11 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:247]: I _think_ that it will not hurt, but in general I would expect people to be so specific in the `URL` that they choose the correct index, but maybe not. 

Continuing

File: common/crypto/platform/openssl/crypto.cpp [hunk 15]
```diff
@@ -50,6 +55,30 @@ using namespace std;
...
+#else
+class OpenSSLResourceHandle {
+public:
+	ProviderPtr default_provider_;
+	ProviderPtr hsm_provider_;
+};
+#endif // MENDER_CRYPTO_OPENSSL_LEGACY
+auto resource_handle_free_func = [](OpenSSLResourceHandle *h) {
+	if (h) {
+		h->~OpenSSLResourceHandle();
```
Comment (kacf) [inline common/crypto/platform/openssl/crypto.cpp:78]: Need to use `delete h` here, to pair it with the `new` below, otherwise the memory of `OpenSSLResourceHandle` itself won't be freed.

But calling `delete` is the default, so actually, you don't need a custom deleter at all.

File: common/crypto/platform/openssl/crypto.cpp [hunk 16]
```diff
@@ -147,11 +182,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
...
+				+ GetOpenSSLErrorMessage()));
+	}
+	log::Debug("Successfully initialised the HSM engine");
+
+	auto private_key = unique_ptr<EVP_PKEY, void (*)(EVP_PKEY *)>(
+		ENGINE_load_private_key(
+			engine.get(),
+			args.private_key_path.c_str(),
+			(UI_METHOD *) NULL,
+			NULL /*callback_data */),
```
Comment (kacf) [inline common/crypto/platform/openssl/crypto.cpp:217]: Should use `nullptr`.

File: common/crypto/platform/openssl/crypto.cpp [hunk 17]
```diff
@@ -147,11 +182,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
...
+
+	// Go through all objects in the context till we find the first private key
+	while (not OSSL_STORE_eof(ctx.get())) {
+		OSSL_STORE_INFO *info = OSSL_STORE_load(ctx.get());
+
+		if (info == NULL) {
+			log::Error(
+				"Failed to load the store info from the hardware security module: trying the next object in the context: "
+				+ GetOpenSSLErrorMessage());
+		}
```
Comment (kacf) [inline common/crypto/platform/openssl/crypto.cpp:272]: Missing `continue`? I see there is one below, but that's after `info` has already been used, so is it correct?

File: common/crypto/platform/openssl/crypto.cpp [hunk 18]
```diff
@@ -147,11 +182,135 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(
...
+			SetupError,
+			"OSSL_STORE_OPEN: Failed to load the private key from the hardware security module: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	// Go through all objects in the context till we find the first private key
+	while (not OSSL_STORE_eof(ctx.get())) {
+		OSSL_STORE_INFO *info = OSSL_STORE_load(ctx.get());
+
+		if (info == NULL) {
```
Comment (kacf) [inline common/crypto/platform/openssl/crypto.cpp:268]: Use `nullptr`.

File: mender-auth/cli/keystore.hpp [hunk 19]
```diff
@@ -68,12 +68,22 @@ class MenderKeyStore {
 	error::Error Save();
 	error::Error Generate();
 
+	string KeyName() {
+		return key_name_;
+	}
+	string SSLEngine() {
+		return ssl_engine_;
+	};
+	string PassPrhase() {
```
Comment (kacf) [inline mender-auth/cli/keystore.hpp:77]: Spelling: `PassPhrase`

File: common/crypto/platform/openssl/crypto.cpp (hunk 15 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:78]: Oh.. true.

But I do need a custom deleter, if I am to use this way of hiding the resource handle definition in the `cpp` file, for otherwise the unique_ptr requires knowledge of the `OpenSSLResourceHandle` struct.

So I guess question is. Is it enough to call delete on `h` ? Or do i also need to explictly call the destructor?

File: common/crypto/platform/openssl/crypto.cpp (hunk 16 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:217]: I've seen that in other work we have used `NULL` when interacting with `C`, so I stuck to that here. 
I can change it, but just so we are clear

File: common/crypto/platform/openssl/crypto.cpp (hunk 17 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:272]: The other continue is for the case where we don't get a `private-key` from the `info` struct, but some other type.

File: common/crypto/platform/openssl/crypto.cpp (hunk 18 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:268]: Same as above

File: common/crypto/platform/openssl/crypto.cpp (hunk 16 above)
Comment (kacf) [reply common/crypto/platform/openssl/crypto.cpp:217]: Right. It's probably because we copied it from C examples or whatever. I'd use `nullptr` since it has an important type safety role.

File: common/crypto/platform/openssl/crypto.cpp (hunk 15 above)
Comment (kacf) [reply common/crypto/platform/openssl/crypto.cpp:78]: Yes, `delete` is enough, it both calls the destructor, and frees the memory.

However, I suspect you will run into the same problem, and I have a hunch why: Since this is a simple object, the destructor is (implicitly) defined in the header, where the full type is not available. Try to declare `~OpenSSLResourceHandle();` in the header, and put an empty one in the cpp file, which I think should be enough to move its definition there, avoiding the need for a full type.

I think...

File: common/crypto/platform/openssl/crypto.cpp (hunk 17 above)
Comment (kacf) [reply common/crypto/platform/openssl/crypto.cpp:272]: Ok, so we do need it here as well?

File: common/crypto/platform/openssl/crypto.cpp (hunk 15 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:78]: Seems `delete` does both :raised_hands:  :)

File: common/crypto/platform/openssl/crypto.cpp (hunk 17 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:272]: Yes :) thank you!

File: common/crypto/platform/openssl/crypto.cpp (hunk 15 above)
Comment (oleorhagen) [reply common/crypto/platform/openssl/crypto.cpp:78]: > However, I suspect you will run into the same problem, and I have a hunch why: Since this is a simple object, the destructor is (implicitly) defined in the header, where the full type is not available. Try to declare ~OpenSSLResourceHandle(); in the header, and put an empty one in the cpp file, which I think should be enough to move its definition there, avoiding the need for a full type.

> I think...

The compiler kept asking for `sizeof(OpenSSLResourceHandle)`, so I don't think we are as lucky :/

**Review by lluiscampos — APPROVED** [review]
It looks good from my side. Nothing more to add :+1:

**Review by kacf — APPROVED** [review]
Thanks!
